pgjdbc/src/main/java/org/postgresql/jdbc/TypeInfoCache.java

```diff
@@ -973,4 +973,17 @@ public boolean requiresQuotingSqlType(int sqlType) throws SQLException {
     }
     return true;
   }
+
+  @Override
+  public int longOidToInt(long oid) {
+    if (oid >= (1L << 32) || oid < 0)
```

Comment:
just do a bitwise comparison with the leading 32 bits

Reply:
something like:

`return (oid & 0xFFFFFFFF_00000000L) == 0 ? (int) oid : 0;`

Reply:
I've updated this code block with the &-trick. 

I opted against a ternary operator, as that would make it a tiny bit less readable, but if  you have strong feelings against the use of an if-statement I'll update it for you.

Reply:
For me, it's actually clearer the ternary operator, but anyway using an if-statement makes sense if you throw an exception which actually would be a safer approach, so ok with using if, but not with returning `0` (zero). 

You are assuming that `0` is an "invalid OID" but it's actually used to represent unspecified Oid https://github.com/pgjdbc/pgjdbc/blob/67fc98821ab6849dc6034393829adab2c64a9baa/pgjdbc/src/main/java/org/postgresql/core/Oid.java#L20

Having an OID that overflows an uint32 should be treated as an error and not blindly ignore the overflow and return it as `0`, it's the case where "it should never happen" but for that reason, an exception should be better.

Reply:
> You are assuming that 0 is an "invalid OID" but it's actually used to represent unspecified Oid 

Hmm, for PostgreSQL itself (https://github.com/postgres/postgres/blob/e56bce5d43789cce95d099554ae9593ada92b3b7/src/include/postgres_ext.h#L33-L37)

```c
#ifdef __cplusplus
#define InvalidOid		(Oid(0))
#else
#define InvalidOid		((Oid) 0)
#endif
```

... which made me think that that value would be handled and called similarly in the JDBC driver.

> Having an OID that overflows an uint32 should be treated as an error and not blindly ignore the overflow and return it as 0, it's the case where "it should never happen" but for that reason, an exception should be better. 

I believe that returning postgres' own "InvalidOid" or PgJDBC's "UNSPECIFIED" is a similarly valid solution. 

Apart from that, would you have a suggestion as to what kind of error to throw, and where to add the translations (if any, assuming that that is what these `GT.tr()`-calls are for)?

Reply:
Maybe something like "Invalid OID returned: {0}"?

The problem with using `UNSPECIFIED` is that PgJDBC makes use of that value with a special meaning, so invalid OIDs could silently be passed by the driver with unexpected results. Anyway, since this "should never happen", probably returning `-1` might be a valid solution too  (since OIDs should not be negative), but I'm not sure about the unexpected behavior of this.

Reply:
> probably returning -1 might be a valid solution too

With this patch we explicitly use the negative range of integers as well to be able to handle the full range of possible OIDs: the first value not representable in java integers (`2**31`) is stored as `-2147483648`, and all subsequent OID values count up from that up to the larges valid OID of `2**32 - 1` which equals `-1` in its java representation. Therefore there are no integer sentinel values left if `UNSPECIFIED` cannot be used. 

Similarly, do note that the String-to-oidInteger parser also wraps to negative numbers when it encounters oids larger than Integer.MAX_INT: https://github.com/pgjdbc/pgjdbc/blob/67fc98821ab6849dc6034393829adab2c64a9baa/pgjdbc/src/main/java/org/postgresql/core/Oid.java#L135

I've updated this code to throw an error `Value is not an OID: {0}` when it encounters a `long` with a value outside the possible range of OIDs.

Reply:
You are right, I forgot the intention of the patch. The error message LGTM, thanks.